timemachine/md/smc.py

```diff
@@ -0,0 +1,159 @@
+# adapted from https://github.com/proteneer/timemachine/blob/7d6099b0f5b4a2d0b26c3edc7a91c18f7a526c00/md/experimental/smc.py
+
+from typing import Any, Callable, Dict, List, Tuple
+
+import numpy as np
+from scipy.special import logsumexp
+from tqdm import tqdm
+from typing_extensions import TypeAlias
+
+# type annotations
+Sample: TypeAlias = Any
+Samples: TypeAlias = List[Sample]  # e.g. List[CoordsVelBox]
+
+Lambda: TypeAlias = float
+LogWeight: TypeAlias = float
+Array: TypeAlias = np.ndarray
+IndexArray: TypeAlias = Array
+LogWeights: TypeAlias = Array
+
+BatchPropagator = Callable[[Samples, Lambda], Samples]
+BatchLogProb = Callable[[Samples, Lambda], LogWeights]
+
+Resampler = Callable[[LogWeights], Tuple[IndexArray, LogWeights]]
+ResultDict = Dict[str, Any]
+
+
+def sequential_monte_carlo(
+    samples: Samples,
+    lambdas: Array,
+    propagate: BatchPropagator,
+    log_prob: BatchLogProb,
+    resample: Resampler,
+) -> ResultDict:
+    """barebones implementation of Sequential Monte Carlo (SMC)
+
+    Parameters
+    ----------
+    samples: [N,] list
+    lambdas: [K,] array
+    propagate: function
+        [move(x, lam) for x in xs]
+        for example, move(x, lam) might mean "run 100 steps of all-atom MD targeting exp(-u(., lam)), initialized at x"
+    log_prob: function
+        [exp(-u(x, lam)) for x in xs]
+    resample: function
+        (optionally) perform resampling given an array of log weights
+
+    Returns
+    -------
+    trajs_dict
+        "sample_traj"
+            [K-1, N] list of snapshots
+        "incremental_log_weights_traj"
+            [K-1, N] array of incremental log weights
+        "ancestry_traj"
+            [K-1, N] array of ancestor idxs
+        "log_weights_traj"
+            [K, N] array of accumulated log weights
+
+    References
+    ----------
+    * Arnaud Doucet's annotated bibliography of SMC
+        https://www.stats.ox.ac.uk/~doucet/smc_resources.html
+    * [Dai, Heng, Jacob, Whiteley, 2020] An invitation to sequential Monte Carlo samplers
+        https://arxiv.org/abs/2007.11936
+    """
+    n = len(samples)
+    log_weights = np.zeros(n)
+
+    # store
+    sample_traj = [samples]
+    ancestry_traj = [np.arange(n)]
+    log_weights_traj = [np.array(log_weights)]
+    incremental_log_weights_traj = []  # note: redundant but convenient
+
+    trange = tqdm(lambdas[:-2])
+
+    def accumulate_results(samples, indices, log_weights, incremental_log_weights):
+        sample_traj.append(samples)
+        ancestry_traj.append(indices)
+        log_weights_traj.append(np.array(log_weights))
+        incremental_log_weights_traj.append(np.array(incremental_log_weights))
+        running_estimate = -logsumexp(log_weights - np.log(len(log_weights)))
+        trange.set_postfix(EXP=running_estimate)
+
+    # main loop
+    for (lam_initial, lam_target) in zip(trange, lambdas[1:-1]):
```

Comment:
Should this be zip(lambdas[:-1], lambdas[1:])?

Reply:
Could be shifted.

Right now it loops from 0 to -2, and then has a block outside the loop to compute importance weights from -2 to -1.

Could rearrange so that the loop goes from 1 to -1, and put a similar block before the loop.

Reply:
This doesn't seem to do a final propagation step though - so the final, end-states samples we gathered are for `lambda[-2]` as opposed to `lambda[-1]`. (Which is problematic for re-weighting since we need equilibrium samples for `lambda[-1]`)

Reply:
>This doesn't seem to do a final propagation step though - so the final, end-states samples we gathered are for lambda[-2] as opposed to lambda[-1].

The final propagation step here targets p(x | lambda[-2]), and then the final weighting step weights for the difference between p(x | lambda[-1]) and p(x | lambda[-2]).

The output of SMC is a weighted sample from p(x | lambda[-1]).

---------

>(Which is problematic for re-weighting since we need equilibrium samples for lambda[-1])

Summarizing from offline discussion: When we want to convert this into a collection of _unweighted_ samples from p(x | lambdas[-1]), we should: (1) resample (e.g. `multinomial_resample`, `stratiified_resample`, `systematic_resample`, ..., to go from weighted to unweighted at the cost of introducing degeneracy), (2) run short propagation targeting p(x | lambdas[-1]) (to mitigate the degeneracy introduced by the last step).

Notes:
* Appending `propagate(xs, lambdas[-1])` to the end of this function won't contribute to the estimate returned by SMC
* Simply appending `propagate(xs, lambdas[-1])` is insufficient to discard the weights
* (subtle point) Appending `resample` and `propagate` is insufficient to discard the weights, since the `resample` function used in the main loop could be optional. (E.g. if the user supplied `null_resample` or `conditional_multinomial_resample`)
* It may be best for this to be handled in application code, or perhaps to add a utility `get_endstate_samples_from_smc_result` to this PR

Reply:
Utility added in https://github.com/proteneer/timemachine/pull/718/commits/ee0616f895e788c06bfcb6b102be059ed79c9df5

Reply:
Utility tested in https://github.com/proteneer/timemachine/pull/718/commits/4b914d94f04a3f1cfabb8da6ceaca8497c59e635